sw/device/lib/crypto/drivers/kmac_test.c

```diff
@@ -61,5 +63,46 @@ bool test_main(void) {
                     current_test_vector->digest_len);
   }
 
+  // Below we repeat some of the test above, but through higher-level
+  // cryptolib API. The gist is to test if glueing is done correctly.
+  for (i = 0; i < nist_kmac_nr_of_vectors; i++) {
+    current_test_vector = nist_kmac_vectors[i];
+
+    crypto_const_uint8_buf_t input_buf;
```

Comment:
I'm curious why we need a separate type for this. If we really want the name of the type to include `const` for some reason, would a `typedef` like `typedef const crypto_uint8_t_buf_t const_crypto_uint8_t_buf_t` work? Being able cast from non-const to const might come handy.

Reply:
The issue is that declaring the struct `const` doesn't make the actual bytes underneath `const` -- it just makes the pointer `const`. A quick code snippet to demonstrate:
```c
  typedef const crypto_uint8_buf_t const_buf_t;

  uint8_t buf[8] = {
    0x01, 0x02, 0x03, 0x04,
    0x05, 0x06, 0x07, 0x08,
  };
  const_buf_t bytes = {
    .data = buf,
    .len = 8
  };
  // Succeeds!
  bytes.data[0] = 0;
  // Fails.
  bytes.data = NULL;
```

The same behavior occurs when I use `const crypto_uint8_buf_t bytes`. But if I use `crypto_const_uint8_buf_t`, then the underlying data is really `const` because it's declared that way in the struct definition.

At the moment, these datatypes are not used at all internally; they're only used at the cryptolib's API interface. So within the cryptolib, we probably won't ever actually call anything with this interface and need to cast to `const`.

Reply:
I was confused because in the current definition of `const_crypto_uint8_t_buf_t`:
* `len` is not `const` even though it's a critical part of the buffer definition
* `data` is a pointer to `const uint8_t`, i.e. you cannot assign to `data[0]` but you can assign to `data`.

I think `typedef const` makes your intent clear enough. I'm not sure how much safety we can really get in C but if you really want to enforce this you probably want a `const` struct with a `const` `len` `field` and a `const` pointer to `const` data `data` field, i.e. `uint8_t const * const data;`.